File: 0496-next-greater-element-i/0496-next-greater-element-i.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        stack<int>st;
        unordered_map<int,int>mp;
        for(auto&i:nums2){
            while(!st.empty() && st.top()<i){
                mp[st.top()]=i;
                st.pop();
            }
            st.push(i);
        }
        vector<int> ans;
        for(auto&i:nums1) ans.push_back(mp.count(i)?mp[i]:-1);
        return ans;
    }
};

File: 0496-next-greater-element-i/0496-next-greater-element-i.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        Deque<Integer> stack = new ArrayDeque<Integer>();
        HashMap<Integer, Integer> map = new HashMap<>();
        for(int index = nums2.length - 1; index >= 0; --index){
            while(!stack.isEmpty() && stack.peek() < nums2[index])
                stack.pop();
            map.put(nums2[index], stack.isEmpty() ? -1 : stack.peek());
            stack.push(nums2[index]);
        }

        int result[] = new int[nums1.length];
        for(int index = 0; index < nums1.length; ++index)
            result[index] = map.get(nums1[index]);
        return result;
    }
}